nano/node/transport/udp.cpp

```diff
@@ -390,12 +421,36 @@ void nano::transport::udp_channels::stop ()
 	std::lock_guard<std::mutex> lock (mutex);
 	local_endpoint = nano::endpoint (boost::asio::ip::address_v6::loopback (), 0);
 
+	// On test-net, close directly to avoid address-reuse issues. On livenet, close
+	// through the strand as multiple IO threads may access the socket.
 	// clang-format off
-	boost::asio::post (strand, [this] {
-		boost::system::error_code ignored;
-		this->socket.close (ignored);
-	});
+	if (node.network_params.network.is_test_network ())
+	{
+		this->close_socket ();
+	}
+	else
+	{
+		boost::asio::dispatch (strand, [this] {
+			this->close_socket ();
+		});
+	}
 	// clang-format on
+
+	// Close all TCP sockets
+	for (auto i (channels.begin ()), j (channels.end ()); i != j; ++i)
```

Comment:
This is more tcp-induced cleanup complexity. This should be inside a tcp_channels class that's dedicated to managing tcp-related things.

Reply:
Updated